### PR TITLE
community: Add Neptune Analytics support to NeptuneGraph

### PR DESCRIPTION
**Pull request by bechbd**
**Description:** Add support for Neptune Analytics to the existing NeptuneGraph object
**Twitter handle:** @bechbd`

### CODE REVIEW

**Review by baskaryan — APPROVED** [review]

File: libs/community/langchain_community/graphs/neptune_graph.py [hunk 1]
```diff
@@ -23,14 +24,15 @@ class NeptuneGraph:
...
+        port: port number for the  Neptune Database instance, default is 8182.  Ignored for Neptune Analytics
+        use_https: whether to use secure connection, default is True.  Ignored for Neptune Analytics
         client: optional boto3 Neptune client
         credentials_profile_name: optional AWS profile name
         region_name: optional AWS region, e.g., us-west-2
-        service: optional service name, default is neptunedata
-        sign: optional, whether to sign the request payload, default is True
+        service: optional service name, default is neptunedata, may also be neptune-graph for Neptune Analytics
+        sign: optional, whether to sign the request payload, default is True.  Ignored for Neptune Analytics
+        graphIdentifier: the graph identifier for a Neptune Analytics graph.  Ignored for Neptune Database
```
Comment (3coins) [inline libs/community/langchain_community/graphs/neptune_graph.py:35]:
```suggestion
        graph_identifier: the graph identifier for a Neptune Analytics graph.  Ignored for Neptune Database
```

File: libs/community/langchain_community/graphs/neptune_graph.py [hunk 2]
```diff
@@ -71,31 +76,46 @@ def __init__(
             else:
                 import boto3
 
+                if service == "neptune-graph":
+                    self.__isNeptuneAnalytics = True
```
Comment (3coins) [inline libs/community/langchain_community/graphs/neptune_graph.py:80]: Should we set this up at the beginning of the block? Also, the rest of the code is using snake case, let’s stick to that.
```python
self.use_analytics_api = service == “neptune-graph”
```

File: libs/community/langchain_community/graphs/neptune_graph.py [hunk 3]
```diff
@@ -71,31 +76,46 @@ def __init__(
...
-                    self.client = session.client(
-                        service,
-                        **client_params,
-                        config=Config(signature_version=UNSIGNED),
-                    )
+                        self.client = session.client(
+                            service,
+                            **client_params,
+                            config=Config(signature_version=UNSIGNED),
+                        )
```
Comment (3coins) [inline libs/community/langchain_community/graphs/neptune_graph.py:118]: We can get rid of the nested block here by returning early.

```python
if service == "neptune-graph":
    self.graph_indentifier = graph_indentifier
    self.client = session.client(service, region_name=region_name)
    return

# remove the else, and continue with the rest of the code..
```

File: libs/community/langchain_community/graphs/neptune_graph.py [hunk 4]
```diff
@@ -104,10 +124,16 @@ def __init__(
...
+                if service == "neptune-graph":
+                    raise ModuleNotFoundError(
+                        "NeptuneGraph requires a boto3 version 1.34.40 or greater."
+                        "Please install it with `pip install -U boto3`."
+                    ) from e
+                else:
+                    raise ModuleNotFoundError(
+                        "NeptuneGraph requires a boto3 version 1.28.38 or greater."
+                        "Please install it with `pip install -U boto3`."
+                    ) from e
```
Comment (3coins) [inline libs/community/langchain_community/graphs/neptune_graph.py:136]: Can simplify this.

```suggestion
                  version = “1.34.40” if service == “neptune_graph” else “1.28.38"                
                  raise ModuleNotFoundError(
                      f"NeptuneGraph requires a boto3 version {version} or greater.”
                      "Please install it with `pip install -U boto3`.”
                  ) from e
```

File: libs/community/langchain_community/graphs/neptune_graph.py [hunk 5]
```diff
@@ -133,7 +159,16 @@ def get_schema(self) -> str:
...
-            return self.client.execute_open_cypher_query(openCypherQuery=query)
+            if self.__isNeptuneAnalytics:
+                resp = self.client.execute_query(
+                    graphIdentifier=self.graphIdentifier,
+                    queryString=query,
+                    parameters=params,
+                    language="OPEN_CYPHER",
+                )
+                return json.loads(resp["payload"].read().decode())
+            else:
```
Comment (3coins) [inline libs/community/langchain_community/graphs/neptune_graph.py:170]: Else is extraneous here.

File: libs/community/langchain_community/graphs/neptune_graph.py [hunk 6]
```diff
@@ -143,21 +178,38 @@ def query(self, query: str, params: dict = {}) -> Dict[str, Any]:
...
+                raise NeptuneQueryException(
+                    {
+                        "message": (
+                            "Summary API is not available for this instance of Neptune,"
+                            "ensure the engine version is >=1.2.1.0"
+                        ),
+                        "details": str(e),
+                    }
+                )
```
Comment (3coins) [inline libs/community/langchain_community/graphs/neptune_graph.py:206]:
```suggestion
            message = "Summary API error occurred on Neptune Analytics” if self.use_analytics_api else (
                "Summary API is not available for this instance of Neptune,”
                "ensure the engine version is >=1.2.1.0”
             ),
            raise NeptuneQueryException(
                {
                    "message": message,
                    "details": str(e),
                }
            ) 
```

File: libs/community/langchain_community/graphs/neptune_graph.py [hunk 7]
```diff
@@ -143,21 +178,38 @@ def query(self, query: str, params: dict = {}) -> Dict[str, Any]:
...
 
         try:
-            summary = response["payload"]["graphSummary"]
+            if self.__isNeptuneAnalytics:
+                summary = response["graphSummary"]
+            else:
+                summary = response["payload"]["graphSummary"]
```
Comment (3coins) [inline libs/community/langchain_community/graphs/neptune_graph.py:212]:
```suggestion
            summary = response["graphSummary”] if self.use_analytics_api else response["payload"]["graphSummary"]
```

**Review by 3coins — CHANGES_REQUESTED** [review]
@bechbd 
Thanks for making these changes. I have added some comments to simplify some of the code. 
A high level comment, should we create a new class perhaps `NeptuneGraphAnalytics` to capture the analytics version? It seems like there are conflicting arguments for analytics vs graph and trying to fit everything into one class might be an overkill. Putting this in another class will also simplify the logic, getting rid of the if else blocks.

@krlawrence
What do you think about putting this in another class? We can abstract out some of the common code to avoid duplication. This is a one-way door, since we don’t want users to switch to a new class later on, so we should make this decision now.

**Comment by bechbd** [comment]
Updated the code to split NeptuneGraph and NeptuneAnalyticsGraph into two separate classes, as it will simplify the maintenance.  I created an abstract base class, NeptuneBaseGraph to store the shared functionality between the two classes.  This did require an update of the NeptuneQAChain to check for the base class instead of the specific class.

**Review by 3coins — COMMENTED** [review]
@bechbd 
Great job on doing the refactoring. Left some suggestions about the base class. Also, I think we can put all classes within the same file `neptune_graph.py`, which will make imports simpler.

**Comment by bechbd** [comment]
Thank you for the feedback @3coins , I've made the changes.

**Comment by 3coins** [comment]
@bechbd 
Code looks good. Can you rebase and fix the CI errors.

**Pull request closed without merging**